### ISSUE TITLE
websteps vs webconnectivity comparison

### ISSUE THREAD

**Issue by bassosimone** [ooni/probe#1797]
We want to perform a comparison of websteps and webconnectivity across these dimensions:

1. size of the JSON produced for each URL;
2. bytes consumed per URL;
3. runtime per URL (accounting for DNS, test helper, TCP connect, TLS handshake, fetching time).

We are not focusing on the ability to precisely flag known censorship cases, for now, because we need to design an algorithm for flagging failures for websteps first. Though, if we collect a JSONL for websteps, we can simulate such an algorithm and we actually have a way to perform also a comparison of precision. But, it seems better to put this specific kind of comparison into another issue, otherwise also this issue here will become huge and long running.

**Comment by bassosimone** [comment]
# Experimental setup

I sketched out Jafar2, which adds support for bandwidth throttling. There is a branch called `jafar2` that contains some experimental code for doing that. The initial design is this:

![experimental-setup](https://user-images.githubusercontent.com/337298/136526031-8eda55e1-1120-4109-8d99-0a918c21c081.png)

I got feedback from @FedericoCeratto and @hellais on how this design could be simplified and improved.

This setup allowed me to choose `netem` and `tbf` parameters to simulate different network conditions that include extra delay, extra losses, and shaping of the incoming/outgoing traffic.

I calibrated this setup by verifying that I could introduce extra delays, losses, and shaping independently.

**Comment by bassosimone** [comment]
# Low-bandwidth environments

One concern when designing websteps has been that of ensuring it's reasonably good in a low bandwidth environment, where reasonably good is not super well defined, but broadly means that it's accurate and still able to work without saturating the bandwidth. There is room for improvement in evaluating OONI in such environments. Also, designing with this case in mind means that you are really using a fraction of the bandwidth in environments with more bandwidth. This seems an optimisation problem. It may be tempting to solve it, but we are also using background measurements a lot, so it may be futile to try and solve this problem. Either way, it's clear that we want to continue measuring and learning here.

Though, in the interest of having some reference data for the future, let me put here some charts with comments taken from a run in the environment that we called 3G above. (In fairness, our emulated environment was not a good emulation of 3G because it did not feature packet batching, but still we have data points in an environment with low bandwidth, which is a good stepping stone for taking initial decisions and running better simulations in the future.)

## Bandwidth usage

The following chart shows the empirical CDF of the download bandwidth usage from the `--bwmon` snapshots.

![bw_down](https://user-images.githubusercontent.com/337298/136531585-d7aeae62-2ec6-4472-8d15-56544673008e.png)

The following chart shows the empirical CDF of the upload bandwidth usage from the `--bwmon` snapshots.

![bw_up](https://user-images.githubusercontent.com/337298/136531752-3e255451-4499-4471-a01d-cd16c883a4bd.png)

The nice part about websteps here is that in these 5s windows we are never very close to the maximum configure download and upload bandwidth (700 kbit/s in this experimental setup).

## Downloading and uploading bytes over time

The following charts are another way of viewing the bandwith usage. This is downloaded data over time:

![bytes_recv](https://user-images.githubusercontent.com/337298/136532244-341ea484-b07a-4151-a2a2-f6a273a31fdf.png)

And this is uploaded data over time:

![bytes_sent](https://user-images.githubusercontent.com/337298/136532286-54faeb9e-7430-4bcf-bb54-0239788265b8.png)

When the line flattens, it means that the probe is idle, because it is either waiting for the test helper or because it is about to take a timeout. (Given the bandwidth constraints, and given how much my home network was faster, I did run some of the measurements using Jafar2 in parallel with a delay between them that varies from few seconds to tens of seconds.)

## Report file size

The following chart shows how the size of the `report.jsonl` file collected on disk incremented during the measurement (which is in itself a proxy of how many bytes we are going to submit).

![report_size](https://user-images.githubusercontent.com/337298/136532030-598503df-b630-478e-b0bf-7f3a7a3d04cb.png)

In other experiments, the difference in the report size was less pronounced. This is one of the cases where websteps has the most advantage, but in other cases they were just quite close.

## Measurement runtime

This chart shows the comparison of the overall measurement runtime:

![measurement_runtime](https://user-images.githubusercontent.com/337298/136532979-bc31292b-4201-4784-af5a-8a2cf4b40d6f.png)

Remember that here "measurement" means "we measure _any_ URL".

Here it's instead the time to perform the DNS step:

![dns_runtime](https://user-images.githubusercontent.com/337298/136533237-41925d78-dc0f-490a-af7b-4e576031dc99.png)

Here it's instead the time to query the respective TH:

![th_runtime](https://user-images.githubusercontent.com/337298/136533287-b96909a7-f58e-4089-90e3-74560429fcf9.png)

Here it's instead the time to complete the measurement (i.e., HTTP-fetch endpoints for websteps, and TCP+TLS+HTTP for webconnectivity):

![epnts_runtime](https://user-images.githubusercontent.com/337298/136533400-7acb39bb-75bc-45fd-89a3-a41acc4676dc.png)

Here we see the trends highlighted above in action.

## Number of measurements emitted over time

Let's start by showing the progress in the number of completed measurements for _any_ URL:

![measurement_start](https://user-images.githubusercontent.com/337298/136533582-4640664b-cdc5-49bd-962e-1f7a364964e4.png)

And let's now instead only consider the progress in emitting _test lists_ URLs (which should be the same for webconnectivity and instead be different for websteps, which makes every redirection explicit in the measurements):

![tlu_start](https://user-images.githubusercontent.com/337298/136533697-26aed46f-6718-4b62-8b26-7a92dedbeb92.png)

**Comment by bassosimone** [comment]
Closing!
